keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+the `resource_id` has already been prepared, the Plugin MUST reply `0
+OK`.
+
+If this RPC failed, or kubelet does not know if it failed or not, it
+MAY choose to call `NodePrepareResource` again, or choose to call
+`NodeUnprepareResource`.
+
+On a successful call this RPC should return set of fully qualified
+CDI device names, which kubelet MUST pass to the runtime through the CRI
+protocol.
+
+```protobuf
+message NodePrepareResourceRequest {
+  // The UID of the ResourceClaim. This field is REQUIRED.
+  string resource_uid = 1;
+}
+
+message NodePrepareResourceResponse {
+  // These are the additional devices that kubelet must
+  // make available via the container runtime. A resource
+  // may have zero or more devices.
+  repeated string cdi_device = 1;
+}
+```
+
+CRI protocol MUST be extended for this purpose, e.g. list of CDI
+device ids should be added to the CRI Device structure:
+
+```protobuf
+// Device specifies a host device to mount into a container.
+message Device {
+    ...
+    string permissions = 3;
+    // Set of fully qualified CDI device names
+    repeated string cdi_device = 4;
+}
+```
+
+###### NodePrepareResource Errors
+
+If the plugin is unable to complete the NodePrepareResource call
+successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST
+return the specified gRPC error code.  Kubelet MUST implement the
+specified error recovery behavior when it encounters the gRPC error
+code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Resource does not exist | 5 NOT_FOUND | Indicates that a resource corresponding to the specified `resource_id` does not exist. | Caller MUST verify that the `resource_id` is correct and that the resource is accessible and has not been deleted before retrying with exponential back off. |
+ 
+
+##### `NodeUnprepareResource`
+
+A Kubelet Plugin MUST implement this RPC call. This RPC is a reverse
+operation of `NodePrepareResource`. This RPC MUST undo the work by
+the corresponding `NodePrepareResource`. This RPC SHALL be called by
+kubelet at least once for each successful `NodePrepareResource`. The
+Plugin SHALL assume that this RPC will be executed on the node where
+the resource is being used.
+
+This RPC is called by kubelet when the Pod using the resource is being
+deleted.
+
+This operation MUST be idempotent. If this RPC failed, or kubelet does
+not know if it failed or not, it can choose to call
+`NodeUnprepareResource` again.
+
+```protobuf
+message NodeUnprepareResourceRequest {
+  // The UID of the ResourceClaim. This field is REQUIRED.
+  string resource_id = 1;
+  // List of fully qualified CDI device names
+  // Kubelet plugin returns them in the NodePrepareResourceResponse
+  repeated string cdi_device = 1;
+}
+
+message NodeUnprepareResourceResponse {
+  // Intentionally empty.
+}
+```
+
+###### NodeUnprepareResource Errors
```

Comment:
what should the kubelet do to the node now?  is the pod stuck terminating?

Reply:
Yes, the pod is stuck. Deleting the pod would incorrectly indicate that the resource is not used by that pod anymore.